src/cdn_definitions/schema.yaml

```diff
@@ -209,6 +208,55 @@ definitions:
       $ref: "#/definitions/absolute_path"
     uniqueItems: true
 
+  iso8601_datetime:
+    type: string
+    anyOf:
+      - format: date
+      - format: date-time
+
+  repo_flag:
+    type: object
+    properties:
+      if_match_id:
+        type: string
+      if_match_path:
+        $ref: "#/definitions/absolute_path"
+      if_created_before:
+        $ref: "#/definitions/iso8601_datetime"
+      if_created_after:
+        $ref: "#/definitions/iso8601_datetime"
+      flag:
+        type: string
+        minLength: 1
+      value:
+        type:
```

Comment:
I'm not sure about allowing value to be something other than a boolean.  I hadn't proposed that. Did you have any particular usage in mind for it?

It's not terrible or anything, I suppose my specific concerns with it would be:

- should it really be called "flag" if it can hold arbitrary values?  I understand the usual meaning of a "flag" in software to be something which is either set or unset.  If we were going to let people store arbitrary strings and numbers on these, we'd probably call it something more like a "field" or an "attribute".
- opens up the possibility for people making more of a mess with the same "flag" having multiple different values, maybe of different types, and no obvious way how those are supposed to be merged or interpreted.  I figured if we support booleans only we can probably have a simple solution like "if flag is explicitly set to both enabled and disabled, then disabled takes precedence" (which is what I've written into AC of RHELDST-17695).  If non-boolean types are allowed it'll probably have to be something more like "if multiple rules match, the last match in the list takes precedence".

Also, no big deal, but if flag values are separated from flag names some minor rewriting of JIRAs will have to be done.  e.g. in RHELDST-17695 I wrote `if the set of active flags contains "exodus" and does not contain "-exodus"`, but with this different approach there may instead be two flags of name "exodus", one with value true and one with value false.

Reply:
Well, the core of the request was _to implement a proper long-term solution to the need to deviate from the baseline config_. I can imagine situations where we need a value different than simple enable / disable. It borderlines on a hypothetical situation, but e.g. the checksum types can be set like that. We use sha256 as the default and we may get a request to start experimenting with sha384, sha512 and sha3. In such case we'd probably want to supply the values directly instead of creating three different mutually exclusive flags.

The idea is also that the rules are evaluated from top to bottom, resp. from first to last (it's a list, after all), and the latest value always takes precedence. In my head this makes the subsequent work in the tools where the flag is needed easier, as it would simply set `repo.flags[flag_name] = flag_value` while iterating over the repos matched by whatever function fetches all the repos (e.g. `rhsm-pulp-manager-v2 --set-repo-distributors`) and then the code actually interested that flag would just do `repo.flags.get(flag_name, default_value)`. On the other hand the YAGNI principle is in favor of sticking to your original idea, because we haven't had the need for such granularity yet. Let me know if you'd feel more comfortable sticking to the original design.

As for the name - I definitely agree there. I envision it more as e.g. kernel parameters, so "parameter", "field" or "attribute" is certainly more appropriate name.

Reply:
Let's go ahead with the way you're proposing.  I would like us to fix the naming though.

Suggestion 1, on each rule, instead of `flag` it could be the generic `key`.  So it's (`<some rule>`, key, value).

Suggestion 2, instead of "repo_flags" one of the following names may be better:

- repo_config
- repo_overrides
- repo_attrs
- repo_attributes